cmd/containerd/server/config/config_test.go

```diff
@@ -248,3 +248,189 @@ func TestDecodePluginInV1Config(t *testing.T) {
+	err := os.WriteFile(filepath.Join(tempDir, "data1.toml"), []byte(data1), 0600)
+	assert.NoError(t, err)
+	err = os.WriteFile(filepath.Join(tempDir, "data2.toml"), []byte(data2), 0600)
+	assert.NoError(t, err)
+
+	// Parse them both
+	var out Config
+	var out2 Config
+	err = LoadConfig(context.Background(), filepath.Join(tempDir, "data1.toml"), &out)
+	assert.NoError(t, err)
+	err = LoadConfig(context.Background(), filepath.Join(tempDir, "data2.toml"), &out2)
+	assert.NoError(t, err)
+
+	// Merge into one config
+	err = mergeConfig(&out, &out2)
+	assert.NoError(t, err)
+
+	// Test if all values are present
+	criPlugin := out.Plugins["io.containerd.grpc.v1.cri"]
+	assert.Equal(t, criPlugin, map[string]interface{}{
+		// originating from first config
+		"cni": map[string]interface{}{
+			"bin_dir": "/cm/local/apps/kubernetes/current/bin/cni",
+		},
+		// originating from second config
+		"registry": map[string]interface{}{
+			"config_path": "/cm/local/apps/containerd/var/etc/certs.d",
+		},
+	})
+}
+
+func TestMergingTwoPluginConfigsMerge(t *testing.T) {
+	// Configuration that customizes the cni bin_dir
+	data1 := `
+[plugins."io.containerd.grpc.v1.cri".cni]
+    bin_dir = "/cm/local/apps/kubernetes/current/bin/cni"
+`
+	// Configuration that customizes the cni conf_dir
+	data2 := `
+[plugins."io.containerd.grpc.v1.cri".cni]
+    conf_dir = "/tmp"
+`
+	// Write both to disk
+	tempDir := t.TempDir()
+	err := os.WriteFile(filepath.Join(tempDir, "data1.toml"), []byte(data1), 0600)
+	assert.NoError(t, err)
+	err = os.WriteFile(filepath.Join(tempDir, "data2.toml"), []byte(data2), 0600)
+	assert.NoError(t, err)
+
+	// Parse them both
+	var out Config
+	var out2 Config
+	err = LoadConfig(context.Background(), filepath.Join(tempDir, "data1.toml"), &out)
+	assert.NoError(t, err)
+	err = LoadConfig(context.Background(), filepath.Join(tempDir, "data2.toml"), &out2)
+	assert.NoError(t, err)
+
+	// Merge into one config
+	err = mergeConfig(&out, &out2)
+	assert.NoError(t, err)
+
+	// Test if all values are present
+	criPlugin := out.Plugins["io.containerd.grpc.v1.cri"]
+	assert.Equal(t, criPlugin, map[string]interface{}{
+		// originating from first config
+		"cni": map[string]interface{}{
+			"conf_dir": "/tmp",
+			// bin_dir is also preserved
+			"bin_dir": "/cm/local/apps/kubernetes/current/bin/cni",
+		},
+	})
+
+	// Restore first config
+	err = LoadConfig(context.Background(), filepath.Join(tempDir, "data1.toml"), &out)
+	assert.NoError(t, err)
+
+	// Test the other way around
+	err = mergeConfig(&out2, &out)
+	assert.NoError(t, err)
+
+	// Test if all values are present
+	criPlugin = out.Plugins["io.containerd.grpc.v1.cri"]
+	assert.Equal(t, criPlugin, map[string]interface{}{
+		// originating from first config
+		"cni": map[string]interface{}{
+			"bin_dir": "/cm/local/apps/kubernetes/current/bin/cni",
+			// conf_dir is also preserved
+			"conf_dir": "/tmp",
+		},
+	})
+}
+
+func TestMergingTwoPluginConfigsRecursively(t *testing.T) {
```

Comment:
Is there a config with an array, possibly array of structs you can include? Those are the cases which hit the more complex logic today.

Reply:
I understand what you mean, I will try to update this PR with that idea (hopefully!) tomorrow. :)
I've been a bit quiet past few days, but it's taking me a bit longer to figure some things out, I've been trying to test a lot of real configurations today, mostly from actual configuration files that we use, and I think I found a few edge cases that were broken after I rebased this PR last week. The unit tests succeed, but I believe they are not testing enough.